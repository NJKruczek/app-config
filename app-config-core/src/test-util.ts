import { join } from 'path';
import { dir } from 'tmp-promise'; // eslint-disable-line import/no-extraneous-dependencies
import { outputFile, remove } from 'fs-extra';
import { stdin } from 'mock-stdin'; // eslint-disable-line import/no-extraneous-dependencies
import { isTestEnvAndShouldNotPrompt } from './logging';

// function that joins the temp dir to a filename
type JoinDir = (filename: string) => string;

export async function withTempFiles(
  files: { [filename: string]: string },
  callback: (inDir: JoinDir, dir: string) => Promise<void>,
): Promise<void>;

export async function withTempFiles(
  files: [string, string][],
  callback: (inDir: JoinDir, dir: string) => Promise<void>,
): Promise<void>;

export async function withTempFiles(
  files: { [filename: string]: string } | [string, string][],
  callback: (inDir: JoinDir, dir: string) => Promise<void>,
) {
  const { path: folder } = await dir();

  try {
    const entries = Array.isArray(files) ? files : Object.entries(files);

    for (const [filename, contents] of entries) {
      await outputFile(join(folder, filename), contents);
    }

    await callback((filename) => join(folder, filename), folder);
  } finally {
    await remove(folder);
  }
}

export async function mockedStdin(
  callback: (send: (text: string) => Promise<void>, end: () => void) => Promise<void>,
) {
  const mock = stdin();
  isTestEnvAndShouldNotPrompt(false);
  process.stdin.isTTY = true;
  process.stdout.isTTY = true;

  // to keep test output clean, just mock out stdout.write
  const origWrite = process.stdout.write;
  const mockWrite = jest.fn();
  process.stdout.write = mockWrite;

  try {
    const send = (text: string) =>
      new Promise<void>((resolve) => {
        setTimeout(() => {
          mock.send(text);
          mock.send('\n');
          resolve();
        }, 0);
      });

    await callback(send, () => mock.end());
  } finally {
    isTestEnvAndShouldNotPrompt(true);
    process.stdin.isTTY = false;
    process.stdout.isTTY = false;
    process.stdout.write = origWrite;
    mock.restore();
  }
}
