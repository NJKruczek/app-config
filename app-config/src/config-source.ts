import { resolve, extname } from 'path';
import { readFile, pathExists } from 'fs-extra';
import { parse as parseTOML, stringify as stringifyTOML } from '@iarna/toml';
import { safeLoad as parseYAML, safeDump as stringifyYAML } from 'js-yaml';
import { parse as parseJSON5, stringify as stringifyJSON5 } from 'json5';
import { Json, JsonObject } from './common';
import { currentEnvironment, defaultAliases, EnvironmentAliases } from './environment';
import { ParsingExtension } from './extensions';
import { ParsedValue } from './parsed-value';
import { AppConfigError, NotFoundError, ParsingError, BadFileType } from './errors';
import { logger } from './logging';

export enum FileType {
  YAML = 'YAML',
  TOML = 'TOML',
  JSON = 'JSON',
  JSON5 = 'JSON5',
}

/** Base class for "sources", which are strategies to read configuration (eg. files, environment variables) */
export abstract class ConfigSource {
  /** Only method that is *required* for all ConfigSources, which is built on in readValue, read, and readToJSON */
  protected abstract readContents(): Promise<[string, FileType]>;

  /** Parses contents of the source */
  async readValue(): Promise<Json> {
    const [contents, fileType] = await this.readContents();

    return parseRawString(contents, fileType);
  }

  /** Reads the contents of the source into a full ParsedValue (not the raw JSON, like readValue) */
  async read(extensions?: ParsingExtension[]): Promise<ParsedValue> {
    const rawValue = await this.readValue();

    return ParsedValue.parse(rawValue, this, extensions);
  }

  /** Ergonomic helper for chaining `source.read(extensions).then(v => v.toJSON())` */
  async readToJSON(extensions?: ParsingExtension[]): Promise<Json> {
    const parsed = await this.read(extensions);

    return parsed.toJSON();
  }
}

/** Read configuration from a single file */
export class FileSource extends ConfigSource {
  public readonly filePath: string;
  public readonly fileType: FileType;

  constructor(filePath: string, fileType?: FileType) {
    super();

    this.filePath = resolve(filePath);
    this.fileType = fileType ?? filePathAssumedType(this.filePath);
  }

  async readContents(): Promise<[string, FileType]> {
    try {
      const content = await readFile(this.filePath);
      logger.verbose(`FileSource read ${this.filePath}`);

      return [content.toString('utf-8'), this.fileType];
    } catch (err) {
      if (err?.code === 'ENOENT') throw new NotFoundError(`File ${this.filePath} not found`);

      throw err;
    }
  }
}

/** Read configuration from a file, found via "glob-like" search (any file format, with support for environment specific files) */
export class FlexibleFileSource extends ConfigSource {
  constructor(
    private readonly filePath: string,
    private readonly environmentOverride?: string,
    private readonly environmentAliases: EnvironmentAliases = defaultAliases,
  ) {
    super();
  }

  private async resolveSource(): Promise<FileSource> {
    const environment = this.environmentOverride ?? currentEnvironment(this.environmentAliases);
    const environmentAlias = Object.entries(this.environmentAliases).find(
      ([, v]) => v === environment,
    )?.[0];

    const filesToTry = [];

    for (const ext of ['yml', 'yaml', 'toml', 'json', 'json5']) {
      if (environment) filesToTry.push(`${this.filePath}.${environment}.${ext}`);
      if (environmentAlias) filesToTry.push(`${this.filePath}.${environmentAlias}.${ext}`);
    }

    // try these after trying environments, which take precedent
    for (const ext of ['yml', 'yaml', 'toml', 'json', 'json5']) {
      filesToTry.push(`${this.filePath}.${ext}`);
    }

    logger.verbose(`FlexibleFileSource is trying to find [${filesToTry.join(', ')}]`);

    for (const filepath of filesToTry) {
      if (await pathExists(filepath)) {
        logger.verbose(`FlexibleFileSource found successful match at ${filepath}`);

        return new FileSource(filepath);
      }
    }

    throw new NotFoundError(
      `FlexibleFileSource could not find file with ${this.filePath}.{yml|yaml|toml|json|json5}`,
    );
  }

  async readContents(): Promise<[string, FileType]> {
    return this.resolveSource().then((source) => source.readContents());
  }

  async read(extensions?: ParsingExtension[]): Promise<ParsedValue> {
    const source = await this.resolveSource();

    return source.read(extensions);
  }
}

/** Read configuration from an environment variable */
export class EnvironmentSource extends ConfigSource {
  constructor(public readonly variableName: string) {
    super();
  }

  async readContents(): Promise<[string, FileType]> {
    const value = process.env[this.variableName];

    if (!value) {
      throw new NotFoundError(`Could not read the environment variable '${this.variableName}'`);
    }

    const inferredFileType = await guessFileType(value);

    logger.verbose(
      `EnvironmentSource guessed that ${this.variableName} is ${inferredFileType} FileType`,
    );

    return [value, inferredFileType];
  }
}

/** Read configuration from a literal JS object */
export class LiteralSource extends ConfigSource {
  constructor(private readonly value: Json) {
    super();
  }

  async readContents(): Promise<[string, FileType]> {
    return [JSON.stringify(this.value), FileType.JSON];
  }

  async readValue(): Promise<Json> {
    return this.value; // overriden just for performance
  }
}

/** Read configuration from many ConfigSources and merge them */
export class CombinedSource extends ConfigSource {
  constructor(public readonly sources: ConfigSource[]) {
    super();

    if (sources.length === 0) {
      throw new AppConfigError('CombinedSource requires at least one source');
    }
  }

  async readContents(): Promise<[string, FileType]> {
    const value = await this.readValue();

    return [JSON.stringify(value), FileType.JSON];
  }

  async readValue(): Promise<Json> {
    return this.readToJSON();
  }

  async read(extensions?: ParsingExtension[]): Promise<ParsedValue> {
    const values = await Promise.all(this.sources.map((source) => source.read(extensions)));

    const merged = values.reduce<ParsedValue | undefined>((acc, parsed) => {
      if (!acc) return parsed;
      return ParsedValue.merge(acc, parsed);
    }, undefined);

    if (!merged) throw new AppConfigError('CombinedSource ended up merging into falsey value');

    Object.assign(merged, { source: this });

    return merged;
  }
}

/** Read configuration from the first ConfigSource that doesn't fail */
export class FallbackSource extends ConfigSource {
  constructor(public readonly sources: ConfigSource[]) {
    super();

    if (sources.length === 0) {
      throw new AppConfigError('FallbackSource requires at least one source');
    }
  }

  // overriden only because it's part of the class signature, normally would never be called
  async readContents(): Promise<[string, FileType]> {
    const value = await this.readValue();

    return [JSON.stringify(value), FileType.JSON];
  }

  // override because readContents uses it (which is backwards from super class)
  async readValue(): Promise<Json> {
    // take the first value that comes back without an error
    for (const source of this.sources) {
      try {
        const value = await source.readValue();
        logger.verbose(`FallbackSource found successful source`);

        return value;
      } catch (error) {
        if (error instanceof NotFoundError) continue;
        throw error;
      }
    }

    throw new NotFoundError('FallbackSource found no valid ConfigSource');
  }

  // override so that ParsedValue is directly from the originating ConfigSource
  async read(extensions?: ParsingExtension[]): Promise<ParsedValue> {
    // take the first value that comes back without an error
    for (const source of this.sources) {
      try {
        const value = await source.read(extensions);
        logger.verbose(`FallbackSource found successful source`);

        return value;
      } catch (error) {
        if (error instanceof NotFoundError) {
          continue;
        }

        throw error;
      }
    }

    throw new NotFoundError('FallbackSource found no valid ConfigSource');
  }
}

export function stringify(config: Json, fileType: FileType, minimal: boolean = false): string {
  switch (fileType) {
    case FileType.JSON:
      return JSON.stringify(config, null, minimal ? 0 : 2);
    case FileType.JSON5:
      return stringifyJSON5(config, null, minimal ? 0 : 2);
    case FileType.TOML:
      return stringifyTOML(config as any);
    case FileType.YAML:
      return stringifyYAML(config);

    default:
      throw new BadFileType(`Unsupported FileType '${fileType as string}'`);
  }
}

export function filePathAssumedType(filePath: string): FileType {
  switch (extname(filePath).toLowerCase().slice(1)) {
    case 'yml':
    case 'yaml':
      return FileType.YAML;
    case 'toml':
      return FileType.TOML;
    case 'json':
      return FileType.JSON;
    case 'json5':
      return FileType.JSON5;
    default:
      throw new BadFileType(
        `The file path "${filePath}" has an ambiguous file type, and a FileType could not be inferred`,
      );
  }
}

export async function parseRawString(contents: string, fileType: FileType): Promise<Json> {
  switch (fileType) {
    case FileType.JSON:
      return JSON.parse(contents) as JsonObject;
    case FileType.YAML:
      return (parseYAML(contents) ?? {}) as JsonObject;
    case FileType.TOML:
      return parseTOML(contents) as JsonObject;
    case FileType.JSON5:
      return parseJSON5(contents) as JsonObject;
    default:
      throw new BadFileType(`Unsupported FileType '${fileType as string}'`);
  }
}

export async function guessFileType(contents: string): Promise<FileType> {
  for (const tryType of [FileType.JSON, FileType.TOML, FileType.JSON5, FileType.YAML]) {
    try {
      await parseRawString(contents, tryType);

      return tryType;
    } catch {
      // parsing errors are expected
    }
  }

  throw new ParsingError(`The provided configuration was not in a detectable/parseable format`);
}
