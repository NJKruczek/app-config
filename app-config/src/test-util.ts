import { join } from 'path';
import { dir } from 'tmp-promise'; // eslint-disable-line import/no-extraneous-dependencies
import { outputFile, remove } from 'fs-extra';

// function that joins the temp dir to a filename
type JoinDir = (filename: string) => string;

export const withTempFiles = async (
  files: { [filename: string]: string },
  callback: (inDir: JoinDir) => Promise<void>,
) => {
  const { path: folder } = await dir();

  try {
    for (const [filename, contents] of Object.entries(files)) {
      await outputFile(join(folder, filename), contents);
    }

    await callback((filename) => join(folder, filename));
  } finally {
    await remove(folder);
  }
};
