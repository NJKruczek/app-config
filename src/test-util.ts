import { join } from 'path';
import { dir } from 'tmp-promise';
import { outputFile, remove } from 'fs-extra';

export const withFakeFiles = async (
  files: [string, string][],
  cb: (dir: string) => Promise<void>,
) => {
  const { path: tmp } = await dir();
  const filenames = files.map(([name]) => join(tmp, name));
  await Promise.all(
    filenames.map(async (name, i) => {
      await outputFile(name, files[i][1]);
    }),
  );

  await cb(tmp);
  await remove(tmp);
};
