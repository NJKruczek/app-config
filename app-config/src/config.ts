import { Json, isObject } from './common';
import { ParsedValue } from './parsed-value';
import { FlexibleFileSource, EnvironmentSource, NotFoundError } from './config-source';
import { defaultExtensions, FileParsingExtension } from './extensions';
import { loadSchema } from './schema';

export interface Configuration {
  /** full configuration plain JSON, with secrets and nonSecrets */
  fullConfig: Json;
  /** parsed configuration value, with metadata (like ConfigSource) still attached */
  parsed: ParsedValue;
  parsedSecrets?: ParsedValue;
  parsedNonSecrets?: ParsedValue;
}

export async function loadConfig(
  fileName = '.app-config',
  environmentVariableName = 'APP_CONFIG',
  environmentOverride?: string,
): Promise<Configuration> {
  // before trying to read .app-config files, we check for the APP_CONFIG environment variable
  const env = new EnvironmentSource(environmentVariableName);

  try {
    const parsed = await env.read();
    return { parsed, fullConfig: parsed.toJSON() };
  } catch (error) {
    // having no APP_CONFIG environment variable is normal, and should fall through to reading files
    if (!(error instanceof NotFoundError)) throw error;
  }

  const [nonSecrets, secrets] = await Promise.all([
    new FlexibleFileSource(fileName, environmentOverride).read(defaultExtensions),
    new FlexibleFileSource(`${fileName}.secrets`, environmentOverride)
      .read(defaultExtensions.concat(markAllValuesAsSecret))
      .catch((error) => {
        // NOTE: secrets are optional, so not finding them is normal
        if (error instanceof NotFoundError) return undefined;

        throw error;
      }),
  ]);

  const parsed = secrets ? ParsedValue.merge(nonSecrets, secrets) : nonSecrets;

  return {
    parsed,
    parsedSecrets: secrets,
    parsedNonSecrets: nonSecrets,
    fullConfig: parsed.toJSON(),
  };
}

export async function loadValidatedConfig(
  fileName = '.app-config',
  environmentVariableName = 'APP_CONFIG',
  environmentOverride?: string,
): Promise<Configuration> {
  const [{ validate }, { fullConfig, parsed, ...rest }] = await Promise.all([
    loadSchema(),
    loadConfig(fileName, environmentVariableName, environmentOverride),
  ]);

  if (!isObject(fullConfig)) throw new Error('Config was not an object');
  validate(fullConfig, parsed);

  return { fullConfig, parsed, ...rest };
}

const markAllValuesAsSecret: FileParsingExtension = (_, value) => () => [
  value,
  { metadata: { fromSecrets: true } },
];
