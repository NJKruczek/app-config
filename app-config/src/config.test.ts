import { join } from 'path';
import { loadConfig, loadValidatedConfig } from './config';
import { FileSource, EnvironmentSource } from './config-source';
import { ReservedKeyError } from './errors';
import { withTempFiles } from './test-util';

describe('Configuration Loading', () => {
  it('loads configuration from a YAML file', async () => {
    await withTempFiles(
      {
        '.app-config.yml': `foo: 42`,
      },
      async (inDir) => {
        const { fullConfig } = await loadConfig({ directory: inDir('.') });

        expect(fullConfig).toEqual({ foo: 42 });
      },
    );
  });

  it('loads configuration from a TOML file', async () => {
    await withTempFiles(
      {
        '.app-config.toml': `foo = 42`,
      },
      async (inDir) => {
        const { fullConfig } = await loadConfig({ directory: inDir('.') });

        expect(fullConfig).toEqual({ foo: 42 });
      },
    );
  });

  it('loads configuration from a JSON file', async () => {
    await withTempFiles(
      {
        '.app-config.json': `{ "foo": 42 }`,
      },
      async (inDir) => {
        const { fullConfig } = await loadConfig({ directory: inDir('.') });

        expect(fullConfig).toEqual({ foo: 42 });
      },
    );
  });

  it('loads configuration from a JSON5 file', async () => {
    await withTempFiles(
      {
        '.app-config.json5': `{ foo: 42 }`,
      },
      async (inDir) => {
        const { fullConfig } = await loadConfig({ directory: inDir('.') });

        expect(fullConfig).toEqual({ foo: 42 });
      },
    );
  });

  it('loads configuration from environment variable', async () => {
    await withTempFiles(
      {
        '.app-config.yml': `foo: 42`,
      },
      async (inDir) => {
        process.env.APP_CONFIG = 'foo: 88';
        const { fullConfig } = await loadConfig({ directory: inDir('.') });

        expect(fullConfig).toEqual({ foo: 88 });
      },
    );
  });

  it('loads environment specific configuration file', async () => {
    await withTempFiles(
      {
        '.app-config.yml': `env: default`,
        '.app-config.dev.yml': `env: development`,
        '.app-config.production.yml': `env: production`,
      },
      async (inDir) => {
        expect((await loadConfig({ directory: inDir('.') })).fullConfig).toEqual({
          env: 'default',
        });

        process.env.NODE_ENV = 'development';
        expect((await loadConfig({ directory: inDir('.') })).fullConfig).toEqual({
          env: 'development',
        });

        process.env.NODE_ENV = 'production';
        expect((await loadConfig({ directory: inDir('.') })).fullConfig).toEqual({
          env: 'production',
        });
      },
    );
  });

  it('loads secrets configuration file', async () => {
    await withTempFiles(
      {
        '.app-config.yml': `base: present`,
        '.app-config.secrets.yml': `secret: present`,
      },
      async (inDir) => {
        expect((await loadConfig({ directory: inDir('.') })).fullConfig).toEqual({
          base: 'present',
          secret: 'present',
        });
      },
    );
  });

  it('treats APP_CONFIG variable as secret', async () => {
    await withTempFiles(
      {
        '.app-config.schema.yml': `
          type: object
          required: [a, b]
          additionalProperties: false
          properties:
            a: { type: boolean, secret: true }
            b: { type: boolean }
        `,
      },
      async (inDir) => {
        process.env.APP_CONFIG = JSON.stringify({ a: true, b: true });
        const { fullConfig } = await loadValidatedConfig({ directory: inDir('.') });
        expect(fullConfig).toEqual({ a: true, b: true });
      },
    );
  });

  it('loads default values and merges them', async () => {
    await withTempFiles(
      {
        '.app-config.yml': `
          a:
            b: true
          d: true
        `,
      },
      async (inDir) => {
        const { fullConfig, parsed } = await loadConfig({
          directory: inDir('.'),
          defaultValues: { a: { c: true }, e: true },
        });

        expect(parsed.sources.find((source) => source instanceof FileSource)).toBeTruthy();
        expect(fullConfig).toEqual({
          a: { b: true, c: true },
          d: true,
          e: true,
        });
      },
    );
  });

  it('loads default values and merges them with APP_CONFIG', async () => {
    process.env.APP_CONFIG = JSON.stringify({
      a: {
        b: true,
      },
      d: true,
    });

    const { fullConfig, parsed } = await loadConfig({ defaultValues: { a: { c: true }, e: true } });

    expect(parsed.sources.find((source) => source instanceof EnvironmentSource)).toBeTruthy();
    expect(fullConfig).toEqual({
      a: { b: true, c: true },
      d: true,
      e: true,
    });
  });

  it('takes default values into account in validation', async () => {
    await withTempFiles(
      {
        '.app-config.yml': `
          a:
            b: true
          d: true
        `,

        '.app-config.schema.yml': `
          type: object
          required: [a, d, e]
          additionalProperties: false
          properties:
            a:
              type: object
              additionalProperties:
                type: boolean
            d: { type: boolean }
            e: { type: boolean }
        `,
      },
      async (inDir) => {
        await expect(loadValidatedConfig({ directory: inDir('.') })).rejects.toThrow();
        await expect(
          loadValidatedConfig({ directory: inDir('.'), defaultValues: { d: 'string' } }),
        ).rejects.toThrow();

        const { fullConfig, parsed } = await loadValidatedConfig({
          directory: inDir('.'),
          defaultValues: { a: { c: true }, e: true },
        });

        expect(parsed.sources.find((source) => source instanceof FileSource)).toBeTruthy();
        expect(fullConfig).toEqual({
          a: { b: true, c: true },
          d: true,
          e: true,
        });
      },
    );
  });
});

describe('Configuration Loading Options', () => {
  it('uses environmentOverride', async () => {
    await withTempFiles(
      {
        '.app-config.yml': `foo: default`,
        '.app-config.prod.yml': `foo: production`,
        '.app-config.development.yml': `
          foo:
            $substitute: '$APP_CONFIG_ENV'
          bar:
            $env:
              default: default
              dev: develop
        `,
      },
      async (inDir) => {
        expect((await loadConfig({ directory: inDir('.') })).fullConfig).toEqual({
          foo: 'default',
        });

        expect(
          (await loadConfig({ directory: inDir('.'), environmentOverride: 'production' }))
            .fullConfig,
        ).toEqual({ foo: 'production' });

        expect(
          (await loadConfig({ directory: inDir('.'), environmentOverride: 'development' }))
            .fullConfig,
        ).toEqual({
          foo: 'development',
          bar: 'develop',
        });
      },
    );
  });
});

describe('CI Environment Variable Extension', () => {
  it('merges APP_CONFIG_EXTEND values', async () => {
    await withTempFiles(
      {
        '.app-config.yml': `foo: 42`,
      },
      async (inDir) => {
        process.env.APP_CONFIG_EXTEND = JSON.stringify({ bar: 88 });
        const { fullConfig } = await loadConfig({ directory: inDir('.') });

        expect(fullConfig).toEqual({ foo: 42, bar: 88 });
      },
    );
  });

  it('merges APP_CONFIG_CI values', async () => {
    await withTempFiles(
      {
        '.app-config.yml': `foo: 42`,
      },
      async (inDir) => {
        process.env.APP_CONFIG_CI = JSON.stringify({ bar: 88 });
        const { fullConfig } = await loadConfig({ directory: inDir('.') });

        expect(fullConfig).toEqual({ foo: 42, bar: 88 });
      },
    );
  });

  it('ignores APP_CONFIG_CI if overriden', async () => {
    await withTempFiles(
      {
        '.app-config.yml': `foo: 42`,
      },
      async (inDir) => {
        process.env.APP_CONFIG_CI = JSON.stringify({ bar: 88 });
        const { fullConfig } = await loadConfig({
          directory: inDir('.'),
          extensionEnvironmentVariableNames: [],
        });

        expect(fullConfig).toEqual({ foo: 42 });
      },
    );
  });
});

describe('V1 Compatibility', () => {
  it('retains nested properties called app-config', async () => {
    await withTempFiles(
      {
        '.app-config.yml': `
          nested:
            app-config:
              extends: base-file.yml
        `,
      },
      async (inDir) => {
        const { fullConfig } = await loadConfig({ directory: inDir('.') });

        // keeps config intact, since app-config isn't at the root
        expect(fullConfig).toEqual({ nested: { 'app-config': { extends: 'base-file.yml' } } });
      },
    );
  });

  it('uses special app-config property for $extends', async () => {
    await withTempFiles(
      {
        '.app-config.yml': `
          app-config: { extends: "base-file.yml" }
          foo: 88
        `,
        'base-file.yml': `
          foo: 42
          bar: foo
        `,
      },
      async (inDir) => {
        const { fullConfig } = await loadConfig({ directory: inDir('.') });

        expect(fullConfig).toEqual({ foo: 88, bar: 'foo' });
      },
    );
  });

  it('uses special app-config property for $extendsOptional', async () => {
    await withTempFiles(
      {
        '.app-config.yml': `
          app-config: { extendsOptional: "base-file.yml" }
          foo: 88
        `,
        'base-file.yml': `
          foo: 42
          bar: foo
        `,
      },
      async (inDir) => {
        const { fullConfig } = await loadConfig({ directory: inDir('.') });

        expect(fullConfig).toEqual({ foo: 88, bar: 'foo' });
      },
    );

    await withTempFiles(
      {
        '.app-config.yml': `
          app-config: { extendsOptional: "base-file.yml" }
          foo: 88
        `,
      },
      async (inDir) => {
        const { fullConfig } = await loadConfig({ directory: inDir('.') });

        expect(fullConfig).toEqual({ foo: 88 });
      },
    );
  });

  it('uses special app-config property for $override', async () => {
    await withTempFiles(
      {
        '.app-config.yml': `
          app-config: { override: "base-file.yml" }
          foo: 88
        `,
        'base-file.yml': `
          foo: 42
          bar: foo
        `,
      },
      async (inDir) => {
        const { fullConfig } = await loadConfig({ directory: inDir('.') });

        expect(fullConfig).toEqual({ foo: 42, bar: 'foo' });
      },
    );
  });

  it('uses special app-config property for $overrideOptional', async () => {
    await withTempFiles(
      {
        '.app-config.yml': `
          app-config: { overrideOptional: "base-file.yml" }
          foo: 88
        `,
        'base-file.yml': `
          foo: 42
          bar: foo
        `,
      },
      async (inDir) => {
        const { fullConfig } = await loadConfig({ directory: inDir('.') });

        expect(fullConfig).toEqual({ foo: 42, bar: 'foo' });
      },
    );

    await withTempFiles(
      {
        '.app-config.yml': `
          app-config: { overrideOptional: "base-file.yml" }
          foo: 88
        `,
      },
      async (inDir) => {
        const { fullConfig } = await loadConfig({ directory: inDir('.') });

        expect(fullConfig).toEqual({ foo: 88 });
      },
    );
  });

  it('uses an ambiguous path in special app-config property', async () => {
    await withTempFiles(
      {
        '.app-config.yml': `
          app-config: { extends: "base-file" }
          foo: 88
        `,
        'base-file.yml': `
          foo: 42
          bar: foo
        `,
      },
      async (inDir) => {
        const { fullConfig } = await loadConfig({ directory: inDir('.') });

        expect(fullConfig).toEqual({ foo: 88, bar: 'foo' });
      },
    );
  });
});

describe('Special values', () => {
  it('fails to loadConfig when a $ prefixed key is seen', async () => {
    process.env.APP_CONFIG = JSON.stringify({ a: { b: { $c: true } } });
    await expect(loadConfig()).rejects.toBeInstanceOf(ReservedKeyError);
  });

  it('loads config when an escaped $ key is seen', async () => {
    process.env.APP_CONFIG = JSON.stringify({ a: { b: { $$c: true } } });
    const { parsed } = await loadConfig();

    expect(parsed.property(['a', 'b'])!.toJSON()).toEqual({ $c: true });
    expect(parsed.property(['a', 'b', '$c'])!.toJSON()).toBe(true);
    expect(parsed.property(['a', 'b', '$c'])!.asPrimitive()).toBe(true);
    expect(parsed.property(['a', 'b', '$c'])!.meta).toMatchObject({ fromEscapedDirective: true });
  });

  it('unescapes $ keys when loading config from file', async () => {
    await withTempFiles(
      {
        '.app-config.yml': `
          a:
            b:
              $$c: true
        `,
      },
      async (inDir) => {
        const { parsed } = await loadConfig({ directory: inDir('.') });

        expect(parsed.property(['a', 'b'])!.toJSON()).toEqual({ $c: true });
        expect(parsed.property(['a', 'b', '$c'])!.toJSON()).toBe(true);
        expect(parsed.property(['a', 'b', '$c'])!.asPrimitive()).toBe(true);
        expect(parsed.property(['a', 'b', '$c'])!.meta).toMatchObject({
          fromEscapedDirective: true,
        });
      },
    );
  });

  it('unescapes double $$ keys when loading config from file', async () => {
    await withTempFiles(
      {
        '.app-config.yml': `
          a:
            b:
              $$$c: true
        `,
      },
      async (inDir) => {
        const { parsed } = await loadConfig({ directory: inDir('.') });

        expect(parsed.property(['a', 'b'])!.toJSON()).toEqual({ $$c: true });
        expect(parsed.property(['a', 'b', '$$c'])!.toJSON()).toBe(true);
        expect(parsed.property(['a', 'b', '$$c'])!.asPrimitive()).toBe(true);
        expect(parsed.property(['a', 'b', '$$c'])!.meta).toMatchObject({
          fromEscapedDirective: true,
        });
      },
    );
  });
});

describe('Dynamic Parsing Extension Loading', () => {
  it('loads a simple extension', async () => {
    await withTempFiles(
      {
        '.app-config.yml': `
          foo: value
        `,
        '.app-config.meta.yml': `
          parsingExtensions:
            - ${join(__dirname, '../../tests/parsing-extensions/uppercase.js')}
        `,
      },
      async (inDir) => {
        const { fullConfig } = await loadConfig({ directory: inDir('.') });

        expect(fullConfig).toEqual({ foo: 'VALUE' });
      },
    );
  });

  it('loads a package as extension', async () => {
    await withTempFiles(
      {
        '.app-config.yml': `
          foo:
            $eval: '2 + 2'
        `,
        '.app-config.meta.yml': `
          parsingExtensions:
            - '@app-config/test-eval-package'
        `,
      },
      async (inDir) => {
        const { fullConfig } = await loadConfig({ directory: inDir('.') });

        expect(fullConfig).toEqual({ foo: 4 });
      },
    );
  });

  it('passes options to package extension', async () => {
    await withTempFiles(
      {
        '.app-config.yml': `
          foo:
            $random: true
        `,
        '.app-config.meta.yml': `
          parsingExtensions:
            - name: '@app-config/test-random-package'
              options:
                seed: consistent
        `,
      },
      async (inDir) => {
        const { fullConfig } = await loadConfig({ directory: inDir('.') });

        expect(fullConfig).toMatchSnapshot();
      },
    );
  });
});

describe('Environment Aliases', () => {
  it('uses environmentAliases as an override', async () => {
    await withTempFiles(
      {
        '.app-config.yml': `
          foo: bar-default
        `,
        '.app-config.production.yml': `
          foo: bar-production
        `,
        '.app-config.meta.yml': `
          environmentAliases:
            Release: production
        `,
      },
      async (inDir) => {
        process.env.APP_CONFIG_ENV = 'Release';
        const { fullConfig } = await loadConfig({ directory: inDir('.') });

        expect(fullConfig).toEqual({ foo: 'bar-production' });
      },
    );
  });

  it('uses environmentAliases in $env', async () => {
    await withTempFiles(
      {
        '.app-config.yml': `
          foo:
            $env:
              default: bar-default
              production: bar-production
        `,
        '.app-config.meta.yml': `
          environmentAliases:
            Release: production
        `,
      },
      async (inDir) => {
        process.env.APP_CONFIG_ENV = 'Release';
        const { fullConfig } = await loadConfig({ directory: inDir('.') });

        expect(fullConfig).toEqual({ foo: 'bar-production' });
      },
    );
  });

  it('uses environmentAliases for $APP_CONFIG_ENV substitution', async () => {
    await withTempFiles(
      {
        '.app-config.yml': `
          foo:
            $subs: bar-$APP_CONFIG_ENV
        `,
        '.app-config.meta.yml': `
          environmentAliases:
            Release: production
          environmentTypeNames:
            - CONFIGURATION
        `,
      },
      async (inDir) => {
        process.env.CONFIGURATION = 'Release';
        const { fullConfig } = await loadConfig({ directory: inDir('.') });

        expect(fullConfig).toEqual({ foo: 'bar-production' });
      },
    );
  });
});

describe('Environment Variable Name', () => {
  it('uses environmentTypeNames', async () => {
    await withTempFiles(
      {
        '.app-config.yml': `
          foo: bar-default
        `,
        '.app-config.production.yml': `
          foo: bar-production
        `,
        '.app-config.meta.yml': `
          environmentTypeNames: CONFIGURATION
        `,
      },
      async (inDir) => {
        process.env.CONFIGURATION = 'production';
        const { fullConfig } = await loadConfig({ directory: inDir('.') });

        expect(fullConfig).toEqual({ foo: 'bar-production' });
      },
    );
  });

  it('uses environmentTypeNames for $env', async () => {
    await withTempFiles(
      {
        '.app-config.yml': `
          foo:
            $env:
              default: bar-default
              production: bar-production
        `,
        '.app-config.meta.yml': `
          environmentTypeNames: CONFIGURATION
        `,
      },
      async (inDir) => {
        process.env.CONFIGURATION = 'production';
        const { fullConfig } = await loadConfig({ directory: inDir('.') });

        expect(fullConfig).toEqual({ foo: 'bar-production' });
      },
    );
  });

  it('uses array for environmentTypeNames', async () => {
    await withTempFiles(
      {
        '.app-config.yml': `
          foo:
            $env:
              default: bar-default
              production: bar-production
        `,
        '.app-config.meta.yml': `
          environmentTypeNames:
            - CONFIGURATION
            - OTHER_ENV
        `,
      },
      async (inDir) => {
        process.env.OTHER_ENV = 'production';
        const { fullConfig } = await loadConfig({ directory: inDir('.') });

        expect(fullConfig).toEqual({ foo: 'bar-production' });
      },
    );
  });

  it('uses environmentTypeNames for $APP_CONFIG_ENV substitution', async () => {
    await withTempFiles(
      {
        '.app-config.yml': `
          foo:
            $subs: $APP_CONFIG_ENV
        `,
        '.app-config.meta.yml': `
          environmentTypeNames:
            - CONFIGURATION
        `,
      },
      async (inDir) => {
        process.env.CONFIGURATION = 'production';
        const { fullConfig } = await loadConfig({ directory: inDir('.') });

        expect(fullConfig).toEqual({ foo: 'production' });
      },
    );
  });
});
