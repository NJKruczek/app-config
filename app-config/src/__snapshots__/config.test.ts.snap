// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dynamic Parsing Extension Loading passes options to package extension 1`] = `
Object {
  "foo": 0.7834231273452891,
}
`;
